klippy/extras/dockable_probe.py

```diff
@@ -0,0 +1,596 @@
+                self.toolhead.manual_move(
+                                        [None, None, self.dock_position[2]],
+                                        self.lift_speed)
+            else:
+                raise self.printer.command_error(
+                    "Cannot attach/detach probe, must home Z axis first")
+
+    # Hop z and return to un-homed state
+    def _force_z_hop(self):
+        this_z = self.toolhead.get_position()[2]
+        if self.last_z != this_z:
+            tpos = self.toolhead.get_position()
+            dock_z = self.dock_position[2] + self.position_endstop
+            self.toolhead.set_position([tpos[0], tpos[1], 0.0, tpos[3]],
+                                homing_axes=[2])
+            self.toolhead.manual_move(
+                [None, None, dock_z],
+                self.lift_speed)
+            kin = self.toolhead.get_kinematics()
+            kin.note_z_not_homed()
+            self.last_z = self.toolhead.get_position()[2]
+
+    # Attaching actions
+    def attach_probe(self, return_pos=None):
+        self.pre_attach_gcode.run_gcode_from_command()
+        if self.get_probe_state() == PROBE_DOCKED:
+            self._align_z()
+        if not return_pos:
+            return_pos = self.init_pos
+        if self._check_distance(dist=self.approach_distance):
+            self._align_to_vector(self.dock_angle)
+        retry = 0
+        while (self.get_probe_state() != PROBE_ATTACHED
+               and retry < self.dock_retries):
+            self._do_attach()
+            retry += 1
+        if self.get_probe_state() != PROBE_ATTACHED:
+            raise self.printer.command_error('Probe attach failed!')
+        self.post_attach_gcode.run_gcode_from_command()
+        if return_pos:
+            if not self._check_distance(return_pos, self.approach_distance):
+                self.toolhead.manual_move(
+                    [return_pos[0], return_pos[1], None],
+                    self.travel_speed)
+    def _do_attach(self):
+        if self.get_probe_state() != PROBE_DOCKED:
+            raise self.printer.command_error(
+                'Attach Probe: Probe not detected in dock, aborting')
+        if self._check_distance(dist=self.approach_distance):
+            self._align_to_vector(self.dock_angle)
+        else:
+            self._move_to_vector(self.dock_angle)
+        self.toolhead.manual_move(
+            [self.dock_position[0],self.dock_position[1], None],
+             self.attach_speed)
+        self.attach_gcode.run_gcode_from_command()
+        if self._check_distance(dist=self.approach_distance):
+            self._align_to_vector(self.dock_angle)
+        self.toolhead.manual_move(
+            [self.approach_position[0], self.approach_position[1], None],
+             self.travel_speed)
+
+    # Detaching actions
+    def detach_probe(self, return_pos=None):
+        self.pre_detach_gcode.run_gcode_from_command()
+        if self.get_probe_state() == PROBE_ATTACHED:
```

Comment:
This part is quite harmful in combination with quad_gantry_level since qgl changes Z height of gantry which can become lower than before, i've actually got probe smashed into bed because of this.